LinearDensity now working with residues/segments

BFedder opened the pull request:
Fixes #3571

LinearDensity did not work with grouping='residues' and grouping='segments'.
This is now fixed, and more tests have been added.

Changes made:
MDAnalysis/analysis/lineardensity.py:
  - in _prepare(self):
    - try/except statement replaced by if statements checking the grouping
      specified for LinearDensity.
    - masses and charges are obtained in a way appropriate for each case.

  - in _single_frame(self):
    - 'positions' variable assignment changed from elem.centroid() to
      elem.atoms.centroid() so it works for residues/segments

  - in _conclude(self):
    - Floating point arithmetic issues could lead to negative radicands. A check
      was implemented to set numbers which are close to zero to precisely zero
      before the square root (for standard deviation) is calculated.

MDAnalysisTests/analysis/test_lineardensity.py:
  - renamed previous test function and added expected masses and charges
  - added test functions for the other groupings with their respective expected
    masses and charges

Regarding my suggested change to _conclude(self): 
I stuck with binsize=5 for the new tests for the other groupings for internal 
consistency in the tests. With this bin size, the positions of the centres of 
geometry are invariant, yielding standard deviations equal to zero. Floating point
imprecision made it possible for negative values on the order of e-13 to e-14 to 
appear and cause a warning from np.sqrt() though, and that is why I added the 
extra step of calculating the radicand and editing it where appropriate beforehand. 
Other solutions are to just ignore this and accept nan values in the result arrays, 
or change the bin size for the tests. I was hesitant to do the later to stay consistent
with the established test, but I'm happy to change this of course if that's the consensus.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

BFedder committed:
LinearDensity now working with residues/segments

LinearDensity did not work with grouping='residues' and grouping='segments'.
This is now fixed, and more tests have been added.

Changes made:
MDAnalysis/analysis/lineardensity.py:
  - in _prepare(self):
    - try/except statement replaced by if statements checking the grouping
      specified for LinearDensity.
    - masses and charges are obtained in a way appropriate for each case.

  - in _single_frame(self):
    - 'positions' variable assignment changed from elem.centroid() to
      elem.atoms.centroid() so it works for residues/segments

  - in _conclude(self):
    - Floating point arithmetic issues could lead to negative radicands. A check
      was implemented to set numbers which are close to zero to precisely zero
      before the square root (for standard deviation) is calculated.

MDAnalysisTests/analysis/test_lineardensity.py:
  - renamed previous test function and added expected masses and charges
  - added test functions for the other groupings with their respective expected
    masses and charges

pep8speaks commented:
Hello @BFedder! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`package/MDAnalysis/analysis/lineardensity.py`](https://github.com/MDAnalysis/mdanalysis/blob/0352033e218e01346409a5891fadcdcb02603d65/package/MDAnalysis/analysis/lineardensity.py):

> [Line 84:80](https://github.com/MDAnalysis/mdanalysis/blob/0352033e218e01346409a5891fadcdcb02603d65/package/MDAnalysis/analysis/lineardensity.py#L84): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 85:80](https://github.com/MDAnalysis/mdanalysis/blob/0352033e218e01346409a5891fadcdcb02603d65/package/MDAnalysis/analysis/lineardensity.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 87:1](https://github.com/MDAnalysis/mdanalysis/blob/0352033e218e01346409a5891fadcdcb02603d65/package/MDAnalysis/analysis/lineardensity.py#L87): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

* In the file [`testsuite/MDAnalysisTests/analysis/test_lineardensity.py`](https://github.com/MDAnalysis/mdanalysis/blob/0352033e218e01346409a5891fadcdcb02603d65/testsuite/MDAnalysisTests/analysis/test_lineardensity.py):

> [Line 79:80](https://github.com/MDAnalysis/mdanalysis/blob/0352033e218e01346409a5891fadcdcb02603d65/testsuite/MDAnalysisTests/analysis/test_lineardensity.py#L79): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 79 characters)



##### Comment last updated at 2022-04-08 21:38:30 UTC

IAlibay commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -146,13 +150,19 @@ def _prepare(self):
-            self.masses = np.array([elem.total_mass() for elem in group])
-            self.charges = np.array([elem.total_charge() for elem in group])
-        except AttributeError:  # much much faster for atoms
+        if self.grouping == "atoms":
```
rather than using 3 `if` it might be worth making this `if/elif/else`? Possibly throwing an error if you have an unknown grouping?

IAlibay commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -146,13 +150,19 @@ def _prepare(self):
             self.masses = self._ags[0].masses
             self.charges = self._ags[0].charges
 
+        if self.grouping in ["residues", "segments"]:
+            self.masses = np.array([elem.atoms.total_mass() for elem in group])
+            self.charges = np.array(
+                [elem.atoms.total_charge() for elem in group])
+
+        if self.grouping == "fragments":
+            self.masses = np.array([elem.total_mass() for elem in group])
+            self.charges = np.array([elem.total_charge() for elem in group])
```
So someone like @richardjgowers is probably going to tell me off for this, but the cost of accessing `.atoms.total_mass()` on an atomgroup is pretty much within error as calling `.total_mass()`.

Would it not make sense here to just use the same logic for residues, segments and fragments?

IAlibay commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -183,6 +194,7 @@ def _single_frame(self):
             key = 'char'
             key_std = 'char_std'
             # histogram for positions weighted on charges
+
```
unnecessary blank line addition?

IAlibay commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -191,6 +203,7 @@ def _single_frame(self):
             self.results[dim][key] += hist
             self.results[dim][key_std] += np.square(hist)
 
+
```
pep8 - single blank line between class methods?

IAlibay reviewed (changes requested):
Couple of quick comments on the first half of the code (didn't get pass _conclude for now)

BFedder commented:
Hi, and thanks for the feedback!
First off, I'll be better about checking PEP8 compliance before opening a PR next time... Sorry! 

I could change the code you mention above to the following to have if/elif/else, raise an exception in the case of unknown groupings, and use the same logic for residues, segments and fragments?

``` python
# Get masses and charges for the selection
if self.grouping == "atoms":
    self.masses = self._ags[0].masses
    self.charges = self._ags[0].charges

elif self.grouping in ["residues", "segments", "fragments"]:
    self.masses = np.array([elem.atoms.total_mass() for elem in group])
    self.charges = np.array(
        [elem.atoms.total_charge() for elem in group])

else:
    raise RuntimeError(
        f"{self.grouping} is not a valid value for grouping.")

```

IAlibay commented:
> First off, I'll be better about checking PEP8 compliance before opening a PR next time... Sorry!

Flake8 can be really useful here, especially if combined with something like VS code. That being said, even of the coredevs are terrible at keeping up with pep8 🙄 

> I could change the code you mention above to the following to have if/elif/else, raise an exception in the case of unknown groupings, and use the same logic for residues, segments and fragments?

Sure, at a quick glance that seems reasonable, although you might want to use an AttributeError instead of a RuntimeError here.

IAlibay commented:
I'm going to cycle the PR (opening and closing it), I'm not sure why gh actions is being silly today...

IAlibay commented:
well at least the webhooks got triggered, that should be enough for now until actions comes back (another day another gh service issue... https://www.githubstatus.com/)

BFedder committed:
changes requested by IAlibay

Changed the if/if/if for reading masses and charges
to if/elif/else.
Will now raise an AttributeError if invalid grouping is chosen.
'residues', 'segments' and 'fragments' now use the same logic
(elem.atoms.total_mass()).

BFedder committed:
Update test_lineardensity.py

BFedder committed:
Update lineardensity.py

BFedder committed:
Update test_lineardensity.py

Exchanged assert_almost_equal with assert_allclose

BFedder committed:
Update CHANGELOG

RMeli commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -199,10 +210,24 @@ def _conclude(self):
+            # appropriate before the square root is calculated.
+            radicand_pos = self.results[dim][
+                'pos_std'] - np.square(self.results[dim]['pos'])
+            for i in range(len(radicand_pos)):
+                if np.isclose(radicand_pos, np.zeros(self.nbins))[i]:
+                    radicand_pos[i] = 0
```
```suggestion
            radicand_pos = np.where(np.isclose(radicand_pos, 0), 0, radicand_pos)
```
or something on those lines.

RMeli replied:
Or even
```suggestion
            radicand_pos[radicand_pos < 0] = 0
```

BFedder replied:
Thanks @RMeli! I was sure there must be a better solution for this and your suggestion looks much cleaner. 
I think using your first suggestion with `np.where` has two advantages: 

- If an actual negative number that's not close to zero shows up at this point, that's an error that should be flagged up/ throw a warning
- positive numbers on the order of e-14 are not meaningful results and should be 0 as well

But your second suggestion looks like it would give better performance. I'm not sure how much of a difference this would make, could people weigh in here? If the first suggestion is not much slower than the second, than that's the one I would go for.

RMeli replied:
Yes, precisely. The first option is safer, the second one is faster. 
```
In [1]: import numpy as np

In [2]: x = np.array([1.0, 1e-10, -1e-10, 0.1])

In [3]: %timeit y = np.where(np.isclose(x, 0), 0, x)
40.4 µs ± 325 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [4]: z = x.copy()

In [5]: %timeit z[z < 0] = 0
2.1 µs ± 17.1 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

```
In [6]: x = np.random.normal(size=10000)

In [7]: %timeit y = np.where(np.isclose(x, 0), 0, x)
90.5 µs ± 1.23 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: z = x.copy()

In [9]: %timeit z[z < 0] = 0
6.72 µs ± 65.2 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

In scientific software, IMHO, correctness is far more important than speed. However, since this is the `_conclude` function, if the code executed before performs enough checks it might be safe to use the second option here (i.e. the only negative values only come from numerical errors).

IAlibay replied:
Because we only have to do this once, I might vote for the safer option here, unless we can guarantee that negative numbers are always floating point issues (I've not really delved into what the cause is here, maybe that is the case).

BFedder replied:
Thanks for this clear breakdown, RMeli! 

IAlibay, I have looked into the code again and I can't see a way for actual negative numbers to appear here. I would suggest the following, then:

`radicand_pos[radicand_pos < 1e-9] = 0`

This is fast and also takes care of floating point errors that give a tiny positive number. 

`x = np.array([1.0, 1e-10, -1e-10, 0.1])`
`%timeit x[x < 1e-9] = 0`
`1.07 µs ± 8.22 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)`

IAlibay replied:
Yeah sure I can live with that

BFedder replied:
Changed now.

orbeckst replied:
Sorry for the drive-by comment: If you're only concerned with negative numbers then I'd keep @RMeli 's suggestion with `x[x<0]` instead of introducing an arbitrary cutoff (and 1e-9 is not necessarily small). Numerical roundoff is normal and if anything I'd want to take care of it at the very end of a calculation and not infer in the middle if something ought to be zero or not. The objective of "no negative numbers" is clear. The objective of "should be zero" is not. Thus I only would implement the first one.

PicoCentauri replied:
I am confused by his operation. Do have an example where this happens? I never had an issue with `nan` in calculating density profiles.

Edit: Actually, I also had them. I [usually](https://gitlab.com/maicos-devel/maicos/-/blob/main/maicos/modules/density.py#L358-360) convert negative number with `np.nan_to_num` to 0 after calculating the standard deviation.

BFedder replied:
That's a good point, @orbeckst, thanks! I have changed this to `x[x<0]` now. 

@PicoCentauri, Ah, I did not know about `np.nan_to_num`, thank you! For now, I stuck with the other suggestion because it is slightly faster, but is there another advantage to nan_to_num?

PicoCentauri replied:
If it is faster it is always better. I am not aware of any other advantages compared to your solution.

RMeli commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -199,10 +210,24 @@ def _conclude(self):
+
+            radicand_char = self.results[dim][
+                'char_std'] - np.square(self.results[dim]['char'])
+            for i in range(len(radicand_char)):
+                if np.isclose(radicand_char, np.zeros(self.nbins))[i]:
+                    radicand_char[i] = 0
```
```suggestion
            radicand_char = np.where(np.isclose(radicand_char, 0), 0, radicand_char)
```
or something on those lines.

RMeli replied:
Or even
```suggestion
            radicand_char[radicand_char < 0] = 0
```

IAlibay committed:
Merge branch 'develop' into lindens

BFedder committed:
Update lineardensity.py

 Updated check of radicands in _conclude(self).

PicoCentauri commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -146,13 +150,19 @@ def _prepare(self):
             self.charges = self._ags[0].charges
 
+        elif self.grouping in ["residues", "segments", "fragments"]:
+            self.masses = np.array([elem.atoms.total_mass() for elem in group])
```
The method `group.total_mass` and `total_charge` have a parameter for `compound` to provide grouping which is faster than list compregensions. Is there a reason not to use it here?

BFedder replied:
No reason that I can see, thanks for this suggestion!

PicoCentauri commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -163,8 +173,9 @@ def _single_frame(self):
-            positions = np.array([elem.centroid() for elem in self.group])
+            # Centre of geometry for residues, segments, fragments
+            positions = np.array(
+                [elem.atoms.centroid() for elem in self.group])
```
The center of geometry has no physical meaning. I would still stick to the center of mass. You can get it with 
`positions  = group.center_of_mass(compound=self.grouping)`

IAlibay replied:
Seems like one of those where the original code thought COM but actually attempted (and failed) to do COG.

I agree with COM here too.

This does need to be clearly signposted in the documentation for this class. Can you see if you could add this somewhere in the docstring, and also add a relevant example of using this for there various groupings too please?

BFedder replied:
Yes, I went with COG because that's what was used previously. I have now changed it to COM and made use of `compound`, thanks. 

@IAlibay, I have changed the class docstring in the Parameters section and in the version history, but I am not sure what a useful example would be, because the user input has not changed and the output for the relevant groupings is only now available, so it should not break any pre-existing scripts. Could you please clarify what examples could help here?

IAlibay replied:
Just the same as what is already under example but saying something along the lines of "if you wanted to do this per residue you could do it this way <code block>". It's a tiny addition but it can be quite a nice thing to read if you're not super sure how the grouping argument works.

BFedder replied:
I have added an example of this now

PicoCentauri commented on testsuite/MDAnalysisTests/analysis/test_lineardensity.py:
```diff
@@ -25,15 +25,70 @@
 
 
-def test_serial():
+def test_grouping_atoms():
```
You may simplify duplicated code with `pytest.mark.parametrize`. Look for examples in the testsuite, there are plenty of them. 

Additionally, there are no tests for the actual calculated density. I would at least add a check for the mean of `results.x.pos` and `results.x.char`

BFedder replied:
I have now refactored the test suite to limit duplicate code, thanks for the suggestion. 

The reason I had not previously added tests for the calculated densities is that I am not sure how I would calculate these on my own. For now, I have taken the output of lineardensity.py and used that as placeholder tests, but that obviously does not do much in the way of actual testing. Do you have some advice on how I could get the "ground truth" data for the densities?

PicoCentauri replied:
You are right. The current values are just magic numbers. These will probably fail when we do something with the code. Depending on how much effort you want to invest I propose the following:

Generate a trajectory based on random positions of particles in a cubic box. Draw the positions from a uniform distribution. Run the analysis and compare the densities. You could do three kinds of system: uncharged particles, charged particles and maybe charged dimers. This is probably beyond the scope of this PR but would be a valuable addition!

BFedder replied:
One reason why I stuck with binsize=5 with my new tests is that the original test, which also used this value, had expected values for `results.x.pos`. I thought that this would at least be partial validation for the restored functionality. But independently determined expected values are definitely still needed here. 

Thanks for the suggestion! For now, I need to focus on my GSoC application/proposal, but I would be interested in working on LinearDensity further. I originally started looking into this module because it looked like a good candidate for more tests on codecov, and that's when I encountered the bug fixed in this PR. It looks like between adding tests for the parallel analysis code, getting expected values for the tests, and issues #2507 and #2508, there is still quite a bit to be done here, and I'd be happy to look into it more after more work on my application. For now, I can open an issue describing the need for more test data, and this PR could get merged?

PicoCentauri replied:
Yes take your time but opening an issue is a good idea!

BFedder replied:
I have now opened issue #3603 for improved tests.

orbeckst replied:
Just a quick comment: "ground truth" tests are excellent when you can get them. However, "regression tests" are the next best thing: they will check that future changes do not change what was once considered working. For this PR I would be totally happy with regression tests, especially as the original code also used them.

Thanks for raising the issue, that's a good approach to doing incremental software development: do small working units and keep track of what needs to be done through issues.

PicoCentauri reviewed (changes requested):
Good work to fix this!

I have some code simplifications and one physical question/addition. For me the center of geometry does not make sense to use as a reference point. It has no physical meaning.

IAlibay commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -146,13 +150,19 @@ def _prepare(self):
+                [elem.atoms.total_charge() for elem in group])
+
+        else:
+            raise AttributeError(
```
Make sure this is covered by tests please.

BFedder committed:
Addressing reviewer comments

* Changed test of radicand to only convert negative numbers to 0
* Now uses CoM rather than CoG for density calculations
* Changed Class docstring to reflect change to CoM
* Now uses parameter for compound instead of list comprehension when
  calculating masses, charges, and CoM
* refactored test_lineardensity.py, make use of pytest functionality
  better
* Added test for case of invalid grouping selection

BFedder committed:
Update lineardensity.py

PicoCentauri reviewed (approved):
Very well. LGTM

BFedder committed:
Update lineardensity.py

Added second example to class docstring

orbeckst commented:
@BFedder can you please resolve the conflicts?

orbeckst commented:
@IAlibay can you please check if your comments were addressed?

BFedder committed:
Merge branch 'develop' into lindens

IAlibay commented on package/MDAnalysis/analysis/lineardensity.py:
```diff
@@ -79,6 +80,8 @@ class LinearDensity(AnalysisBase):
        ldens.run()
        print(ldens.results.x.pos)
 
+       # alternatively, specify grouping/binsize like this
+       ldens = LinearDensity(selection, grouping='residues', binsize=1.0)
```
```suggestion

    Alternatively, the other types of groupings can be selected using the ``grouping``
    keyword. For example to calculated the density based on the :class:`ResidueGroup`s
    of the system:
    
    .. code-block:: python
       ldens = LinearDensity(selection, grouping='residues', binsize=1.0)
       ldens.run()

```

IAlibay commented on testsuite/MDAnalysisTests/analysis/test_lineardensity.py:
```diff
@@ -22,18 +22,91 @@
+expected_xchar_fragments = np.array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
+
+
+@pytest.mark.parametrize("grouping, expected_masses, expected_charges,\
+                         expected_xpos, expected_xchar", [
+                         ("atoms",
+                          expected_masses_atoms,
+                          expected_charges_atoms,
+                          expected_xpos_atoms,
+                          expected_xchar_atoms),
+                         ("residues",
+                          expected_masses_residues,
+                          expected_charges_residues,
+                          expected_xpos_residues,
+                          expected_xchar_residues),
+                         ("segments",
+                          expected_masses_segments,
+                          expected_charges_segments,
+                          expected_xpos_segments,
+                          expected_xchar_segments),
+                         ("fragments",
+                          expected_masses_fragments,
+                          expected_charges_fragments,
+                          expected_xpos_fragments,
+                          expected_xchar_fragments)
+                         ])
```
PEP8 is a suggestion not an obligation - don't use this against me

```suggestion
@pytest.mark.parametrize("grouping, expected_masses, expected_charges, expected_xpos, expected_xchar", [
    ("atoms", expected_masses_atoms, expected_charges_atoms,
     expected_xpos_atoms, expected_xchar_atoms),
    ("residues", expected_masses_residues, expected_charges_residues,
     expected_xpos_residues, expected_xchar_residues),
    ("segments", expected_masses_segments, expected_charges_segments,
     expected_xpos_segments, expected_xchar_segments),
    ("fragments", expected_masses_fragments, expected_charges_fragments,
     expected_xpos_fragments, expected_xchar_fragments)
])
```

IAlibay reviewed (changes requested):
Just two small suggested changes - I'm directly suggesting them here since they are more "my style" than anything. Feel free to say no, I'll approve after.

BFedder committed:
Update package/MDAnalysis/analysis/lineardensity.py

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

BFedder committed:
Update testsuite/MDAnalysisTests/analysis/test_lineardensity.py

Co-authored-by: Irfan Alibay <IAlibay@users.noreply.github.com>

BFedder commented:
Thanks, I have taken the suggestions on board

IAlibay reviewed (approved):
Thanks @BFedder 

@MDAnalysis/gsoc-mentors - I can't merge this as a GSoC PR due to conflicts of interest. Can someone please take over the responsibility of merging this?

orbeckst commented:
@PicoCentauri can you please have a final look and then squash-merge? Thanks.

Commit:
Merge branch 'develop' into lindens

PicoCentauri commented:
Thanks @BFedder for your fix!
